src/Symfony/Component/Messenger/Bridge/Kafka/Transport/KafkaReceiver.php

```diff
@@ -0,0 +1,151 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace Symfony\Component\Messenger\Bridge\Kafka\Transport;
+
+use Psr\Log\LoggerInterface;
+use RdKafka\KafkaConsumer;
+use Symfony\Component\Messenger\Envelope;
+use Symfony\Component\Messenger\Exception\TransportException;
+use Symfony\Component\Messenger\Transport\Receiver\ReceiverInterface;
+use Symfony\Component\Messenger\Transport\Serialization\SerializerInterface;
+
+/**
+ * @author Konstantin Scheumann <konstantin@konstantin.codes>
+ */
+class KafkaReceiver implements ReceiverInterface
+{
+    /** @var LoggerInterface */
+    private $logger;
+
+    /** @var SerializerInterface */
+    private $serializer;
+
+    /** @var RdKafkaFactory */
+    private $rdKafkaFactory;
+
+    /** @var KafkaReceiverProperties */
+    private $properties;
+
+    /** @var KafkaConsumer */
+    private $consumer;
+
+    /** @var bool */
+    private $subscribed;
+
+    public function __construct(
+        LoggerInterface $logger,
+        SerializerInterface $serializer,
+        RdKafkaFactory $rdKafkaFactory,
+        KafkaReceiverProperties $properties
+    ) {
+        $this->logger = $logger;
+        $this->serializer = $serializer;
+        $this->rdKafkaFactory = $rdKafkaFactory;
+        $this->properties = $properties;
+
+        $this->subscribed = false;
+    }
+
+    public function get(): iterable
+    {
+        $message = $this->getSubscribedConsumer()->consume($this->properties->getReceiveTimeoutMs());
+
+        switch ($message->err) {
+            case \RD_KAFKA_RESP_ERR_NO_ERROR:
+                $this->logger->info(sprintf(
+                    'Kafka: Message %s %s %s received ',
+                    $message->topic_name,
+                    $message->partition,
+                    $message->offset
+                ));
+
+                $envelope = $this->serializer->decode([
+                    'body' => $message->payload,
+                    'headers' => $message->headers,
+                    'key' => $message->key,
+                    'partition' => $message->partition,
+                    'offset' => $message->offset,
+                    'timestamp' => $message->timestamp,
```

Comment:
Looks wrong to me. Most of serializer only handle `body` and `header` keys

Reply:
I understand why it looks uncommon. But these fields are really valuable for many Teams implementing Kafka.
For example the `timestamp` is about event time. It falls back to "ingestion time" of the message when not set in the producer but commonly it denotes the actual time when the event had occurred.

The `key` is the crucial datum used for compacting the topic. So it is an essential component of implementing Kafka. Also, many Applications communicate the ID of the message / Domain object via this field.

For the `partition`/`offset` combination: it is typically less relevant at the receiving end to have this data. Only in special use cases that I can imagine, would you need to know it.
* `partition`: When the Team decides that the partition denotes a specific kind of message or key range. That would imply that the producing side implemented a custom partitioner. This is not very common but often enough to be relevant.
* `offset`: Probably not very important to the application, except if the processing depends on the order of messages. For example if you create a topic with 1 partition and produce to it using transactions. Then Kafka gives you an ordering guarantee. The Symfony Application might want to check ordering and cache unordered messages into a cache or implement some other security mechanism ensuring messages are processed in an ordered fashion.

Reply:
I think you missed my point. Look at the code of `PhpSerializer`, You'll notice that everything but BODY and HEADER will be ignored.

Reply:
That is of course correct. And if they are ignored, that is not a problem at all, if there are no other applications using the kafka topic.

But as soon as you actually use Kafka the way it is typically used (with a Schema like AVRO + Schema Registry), then you have to provide a custom Serializer in any case.

`PhpSerializer` will only work, if you use Kafka as a simple message queue which is not connected to any other app.